### PR TITLE
HIVE-26840 : Branch 3 netty upgrade to 4.1.69.Final.

### PR DESCRIPTION
**Pull request by amanraj2520**
### What changes were proposed in this pull request?

Changes were made to shade jetty jars that were causing test failures. Previously in branch-3.1 when encountered this error, we reverted it back. But now for the new release we want to use the latest netty version so that we get rid of CVEs.

We found that a lot of tests failed due to incompatibility of Jackson 2.12.0 and Arrow version 0.8.0. We had 2 options either to downgrade jackson to 2.10.0 or upgrade arrow to 0.11.0. We went ahead with upgrading arrow to 0.11.0 since it fixed a lot of CVE's as well. This is fixed in https://github.com/apache/hive/pull/3902. This also brought an issue with test org.apache.hadoop.hive.ql.io.arrow.TestArrowColumnarBatchSerDe which is fixed in [HIVE-26892](https://issues.apache.org/jira/browse/HIVE-26892) / https://github.com/apache/hive/pull/3898 by @cnauroth 

### Why are the changes needed?

For the test cases in the build to pass for branch-3.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Have tested it locally by building Hive branch-3 and the previously failed test cases passed. Please note that only the test cases failing due to netty upgrade were fixed. More test cases fixes need to go in. So this PR will not be successful only test cases with netty failures will be fixed. To ensure that the netty test cases are fixed, the number of test case failure should match the number of test case failures when I reverted this netty version. Jenkins build when reverted the netty version : http://ci.hive.apache.org/blue/organizations/jenkins/hive-precommit/detail/PR-3850/1/pipeline

### CODE REVIEW

**Comment by amanraj2520** [comment]
@cnauroth Can you please review this PR. I have fixed the netty issues as suggested by you in this PR : https://github.com/apache/hive/pull/3850. I have mentioned how we are going to ensure that this change works in the description.

**Review by cnauroth — APPROVED** [review]
+1 (non-binding) pending CI run without test failures related to Netty (though potentially other test failures that still need to be addressed outside this patch).

Thanks for taking care of this, @amanraj2520 !

**Comment by amanraj2520** [comment]
Hi,

The tests have ran and we have the same number of failures as compared to when we revreted the netty version. Adding the snapshots of the number of test failure when we reverted this netty version.
image

This implies that upgrading the netty version works and we can merge this PR. @zabetak @ayushtkn @abstractdog Can either one of you approve and merge this PR 
cc - @cnauroth

**Comment by abstractdog** [comment]
I'm a bit confused with this PR (sorry, I think I missed some context): originally, Jira says "Backport of [HIVE-23073](https://issues.apache.org/jira/browse/HIVE-23073) and [HIVE-24138](https://issues.apache.org/jira/browse/HIVE-24138)", where HIVE-24138 is clearly about removing the relocation for instance, and I cannot see it now in this PR, so what's the scope here?

**Comment by amanraj2520** [comment]
Hi @abstractdog Can you please refer to the conversation in this PR : https://github.com/apache/hive/pull/3850. Changes were made to shade jetty jars that were causing test failures. Previously in branch-3.1 when encountered this error, we reverted it back. But now for the new release we want to use the latest netty version so that we get rid of CVEs.

**Comment by abstractdog** [comment]
but if we keep shading netty jars after this PR then we're simply not backporting [HIVE-24138](https://issues.apache.org/jira/browse/HIVE-24138), right? just asking for clarity's sake

**Comment by amanraj2520** [comment]
I have cherry picked both [HIVE-23073](https://issues.apache.org/jira/browse/HIVE-23073) and [HIVE-24138](https://issues.apache.org/jira/browse/HIVE-24138) since I was of the impression that they were related. Some older netty jars were coming as transitive dependencies thats why https://issues.apache.org/jira/browse/HIVE-24138 was created. Please correct me if I am wrong

**Comment by abstractdog** [comment]
they are related and HIVE-24138 fixed issues caused by [HIVE-23073](https://issues.apache.org/jira/browse/HIVE-23073), my concern is that I can see differences:

original HIVE-24138 commit:
https://github.com/apache/hive/commit/c345cdf58ddf212c2f2d6b82e57ed3d41647ffa7

your backport on this PR (cannot see the ql/pom.xml changes)
https://github.com/apache/hive/pull/3859/commits/d3d2aadb7999b7bcd0d0a96fe79a60665417d2c3

**Comment by amanraj2520** [comment]
Let me check that. Will re-raise this PR. Thanks @abstractdog

**Comment by amanraj2520** [comment]
Hi @abstractdog Sorry my bad, somehow these 2 changes got ignored by me. I have included those now. @abstractdog Can you please review once again

**Comment by amanraj2520** [comment]
@abstractdog Just a gentle reminder, can you please review this.

File: pom.xml
Comment (amanraj2520) [reply pom.xml:191]: Thats a good point. Sure will make this change and test it.

**Review by abstractdog — APPROVED** [review]
+1
pending tests

**Comment by abstractdog** [comment]
there are 10 new failures, can you please check if they are related?
http://ci.hive.apache.org/blue/organizations/jenkins/hive-precommit/detail/PR-3859/4/

**Comment by amanraj2520** [comment]
Sure let me check. Thanks for highlighting

**Comment by amanraj2520** [comment]
@abstractdog The tests start failing when I upgrade netty to 4.1.69 with this error :
ava.util.concurrent.TimeoutException
	at io.netty.util.concurrent.DefaultPromise.get(DefaultPromise.java:359)
	at org.apache.hive.spark.client.TestSparkClient$4.call(TestSparkClient.java:158)
	at org.apache.hive.spark.client.TestSparkClient.runTest(TestSparkClient.java:307)
	at org.apache.hive.spark.client.TestSparkClient.testSyncRpc(TestSparkClient.java:154)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:47)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:44)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:271)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:70)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:50)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:238)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:63)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:236)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:53)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:229)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:309)
	at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:365)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeWithRerun(JUnit4Provider.java:273)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:238)
	at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:159)
	at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:379)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:340)
	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:125)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:413)

If we keep the version as 4.1.48 it passes. Let me investigate on this from my side. Thanks.

**Comment by amanraj2520** [comment]
@abstractdog There is a test org.apache.hadoop.hive.ql.io.arrow.TestArrowColumnarBatchSerDe which is failing with this error :
`java.lang.NullPointerException: src
	at io.netty.util.internal.ObjectUtil.checkNotNull(ObjectUtil.java:39)
	at io.netty.buffer.UnsafeByteBufUtil.setBytes(UnsafeByteBufUtil.java:537)
	at io.netty.buffer.PooledUnsafeDirectByteBuf.setBytes(PooledUnsafeDirectByteBuf.java:199)
	at io.netty.buffer.WrappedByteBuf.setBytes(WrappedByteBuf.java:486)
	at io.netty.buffer.UnsafeDirectLittleEndian.setBytes(UnsafeDirectLittleEndian.java:34)
	at io.netty.buffer.ArrowBuf.setBytes(ArrowBuf.java:937)
	at org.apache.arrow.vector.BaseVariableWidthVector.setBytes(BaseVariableWidthVector.java:1155)
	at org.apache.arrow.vector.BaseVariableWidthVector.setSafe(BaseVariableWidthVector.java:990)
	at org.apache.hadoop.hive.ql.io.arrow.Serializer.writePrimitive(Serializer.java:418)
	at org.apache.hadoop.hive.ql.io.arrow.Serializer.write(Serializer.java:213)
	at org.apache.hadoop.hive.ql.io.arrow.Serializer.writeStruct(Serializer.java:279)
	at org.apache.hadoop.hive.ql.io.arrow.Serializer.write(Serializer.java:219)
	at org.apache.hadoop.hive.ql.io.arrow.Serializer.serializeBatch(Serializer.java:142)
	at org.apache.hadoop.hive.ql.io.arrow.Serializer.serialize(Serializer.java:528)
	at org.apache.hadoop.hive.ql.io.arrow.ArrowColumnarBatchSerDe.serialize(ArrowColumnarBatchSerDe.java:260)
	at org.apache.hadoop.hive.ql.io.arrow.TestArrowColumnarBatchSerDe.serializeAndDeserialize(TestArrowColumnarBatchSerDe.java:240)
	at org.apache.hadoop.hive.ql.io.arrow.TestArrowColumnarBatchSerDe.initAndSerializeAndDeserialize(TestArrowColumnarBatchSerDe.java:204)
	at org.apache.hadoop.hive.ql.io.arrow.TestArrowColumnarBatchSerDe.testComprehensive(TestArrowColumnarBatchSerDe.java:381)`

This is because arrow 0.x had io.netty version of 4.1.22.Final which is the old version. When we try to upgrade it to 4.1.69.Final netty and exclude the netty jar from arrow dependency as I have done in this PR, then we get the above exception in the test. The solutions are as follows :
1. Fix this test with the current netty version. I am still working on it but till now not found a fix. If you know of something please let me know.
2. Revert the netty version. This test works perfectly fine if I change the netty version to 4.1.48.Final.
3. Upgrade to arrow version 2.0.0 as done in master which has its own implementation of netty classes but not sure how many cherry picks will be required for a green build.

**Comment by amanraj2520** [comment]
@ayushtkn @cnauroth @zabetak Adding you for your suggestions. Kindly let me know.

**Comment by cnauroth** [comment]
Hello @amanraj2520 . Referring to the options above:

1. I think the ideal path is to get the test fixed. See below for more analysis from me and a proposed path forward.
2. I'd prefer not to revert back to the earlier version, because we set a goal on the 3.2 release to upgrade dependencies and clear out their CVEs as much as possible. That said, I reviewed CVEs on the older release, and I don't think they have much practical impact on Hive, so I'm not opposed to this as a fallback option if we get stuck.
3. I'd be concerned about a significant major version bump all the way to Arrow 2.0.0. I don't know Arrow well enough to comment on backward-compatibility of that upgrade.

The test is failing specifically on [serializing a row of nulls in all columns](https://github.com/apache/hive/blob/branch-3/ql/src/test/org/apache/hadoop/hive/ql/io/arrow/TestArrowColumnarBatchSerDe.java#L374-L378). I confirmed that it's this row specifically by commenting out the row and seeing the test pass. I also confirmed that it's specifically failing while serializing a struct column. (Null values for primitive types are fine.)

The problem appears to be that the serializer does not correctly track null values within a null struct. We should be calling the Arrow `vector.setNull`, but instead, it ends up calling `vector.setSafe`. There is another patch, [HIVE-25243](https://issues.apache.org/jira/browse/HIVE-25243) / PR #2391 that fixed this on master. I tried applying both your patch and a slightly different version of HIVE-25243, and then the test passed locally. The only thing I don't understand is why this was ever passing with the old version. I guess there are some versions of Arrow + Netty that are just more tolerant of clients calling `vector.setSafe` with null values.

I propose that first, we merge in the current pull request, even with a known test failure. There are already a lot of changes in this patch. Then, I can queue up a separate backport of HIVE-25243. This is non-binding though, so let's see if we can get confirmation on the plan from a committer.

**Comment by cnauroth** [comment]
I filed [HIVE-26892](https://issues.apache.org/jira/browse/HIVE-26892) / #3898 to backport handling of values in null structs. I'll defer to committers on how they'd best want to handle sequencing of merging these patches into branch-3. @amanraj2520 , I hope this helps.

**Comment by amanraj2520** [comment]
Hi @cnauroth Thanks a lot for your analysis. This helps me a lot. I will check this from my side in sometime and let you know. The thing that you were curious about, i think I can help sharing some light. Arrow started using netty in its 0.x releases. If we do not exclude netty from arrow while importing dependencies it will get the 4.1.22.Final netty. A point to be noted is that 4.1.22.Final and 4.1.28.Final(which I was planning to revert to) does not have this nonNull check in its stack trace. Therefore even if we upgrade netty to 4.1.69.Final and keep arrow as 0.x, it will pickup the 4.1.22.Final which will not have the nonNull check and it will pass. The problem came because we removed the arrow's netty in branch-3 while importing therefore forcing arrow to use 4.1.69.Final which has this null check and therefore this test fails. Hope this answers you question. We can discuss it further if you want. 

Having said this i also feel that fixing this test is the best approach and I will try fixing that in another pr. Will wait for someone to approve this first. Thanks a lot @cnauroth

**Comment by amanraj2520** [comment]
@cnauroth Your proposed fix works on my local as well. So according to me we can merge this first and then merge the PR you have raised [HIVE-26892](https://issues.apache.org/jira/browse/HIVE-26892) / https://github.com/apache/hive/pull/3898

I will also change the name of my PR to Upgrade netty to 4.1.69.Final and arrow to 0.11.0.
@abstractdog Can you please review this idea and let me know if you have any concerns. Thanks.

**Comment by abstractdog** [comment]
is this whole this working by adding [HIVE-26892](https://issues.apache.org/jira/browse/HIVE-26892) + having Netty 4.1.69.Final and arrow to 0.11.0?
if so, is there a chance to backport arrow upgrade separately first if it was done separately on master?

**Comment by amanraj2520** [comment]
@abstractdog Yes me and @cnauroth have verified it on our local machine by merging this and the [HIVE-26892](https://issues.apache.org/jira/browse/HIVE-26892) branch. The tests pass. I agree that we can upgrade arrow first on branch-3 and then re-visit this PR only for netty. I had just raised it in the same branch, so that we could establish the fact that arrow and netty will have to go in for a green branch-3. Since now you are aware of the test failures, it would became easier to explain. Will raise the arrow upgrade PR in sometime and let you know.

**Comment by amanraj2520** [comment]
Raised this PR to track the arrow upgrade https://github.com/apache/hive/pull/3902

cc @abstractdog @cnauroth

**Comment by amanraj2520** [comment]
The arrow upgrade is now handled in branch-3 through this PR : https://github.com/apache/hive/pull/3902. Raising this PR again only for Jetty Upgrade @cnauroth @abstractdog @zabetak FYI

**Comment by amanraj2520** [comment]
@abstractdog Can you please merge this PR.

**Review by cnauroth — APPROVED** [review]
+1 (non-binding)

Thank you for sticking with this, @amanraj2520 !

**Comment by amanraj2520** [comment]
Hi @abstractdog @zabetak Can you please review and merge this PR. The tests which are failing are flaky. I have raised separate PR's to disable them as they were disabled on the master as well. Here are the PR's : https://github.com/apache/hive/pull/3929 and https://github.com/apache/hive/pull/3928

**Comment by amanraj2520** [comment]
@abstractdog @zabtek Can you please check this PR and merge if it looks good.

**Comment by amanraj2520** [comment]
@abstractdog A gentle reminder. Can you please merge this PR. It has already been approved.

**Comment by amanraj2520** [comment]
@abstractdog Really apologize to bother you so many times for this PR. The reason why I am stressing on this PR so many times is because this needs to go in branch-3 so that other tests failures (if any) can come up on this branch. For example, this issue prevents tests like CliDriver etc not to run properly. This means that a lot of tests do not run until this is merged to branch-3. So, other PR's are blocked until this PR goes in. 

Requesting you to review and merge this. Sorry for bothering you again.

**Comment by amanraj2520** [comment]
Thanks a lot @abstractdog for merging this

**Pull request merged**
